QUESTION 1:

SOLUTION:  1. DDL (Data Definition Language)
Purpose: DDL commands are used to define, modify, and manage the structure of database objects (like tables, indexes, users, etc.). They deal with the schema or blueprint of the database.
Impact: Changes are permanent and affect the database structure itself.
Examples: CREATE, ALTER, DROP, TRUNCATE, RENAME.

Example SQL:
CREATE TABLE Employees (
    EmployeeID INT PRIMARY KEY,
    FirstName VARCHAR(50),
    LastName VARCHAR(50),
    HireDate DATE
);

2. DML (Data Manipulation Language)
Purpose: DML commands are used to manage data within the schema objects. They deal with adding, retrieving, modifying, or deleting data records in a database.
Impact: Changes affect the data stored in the tables.
Examples: INSERT, UPDATE, DELETE, MERGE.

Example SQL:
INSERT INTO Employees (EmployeeID, FirstName, LastName, HireDate)
VALUES (1, 'John', 'Doe', '2023-01-15');

3. DQL (Data Query Language)
Purpose: DQL commands are used for retrieving data from the database. Technically, SELECT is often considered part of DML, but it's frequently categorized separately as DQL due to its distinct purpose of data retrieval rather than manipulation.
Impact: Does not change the database structure or data; it only reads information.
Examples: SELECT.

Example SQL:
SELECT FirstName, LastName FROM Employees WHERE HireDate > '2023-01-01';

QUESTION 2:

SOLUTION: Purpose of SQL Constraints:
Data Integrity: They ensure that the data stored in the database adheres to specific business rules and is consistent.
Data Accuracy: They prevent the entry of invalid data, thereby improving the accuracy of the information.
Data Reliability: By enforcing rules, they make the data more dependable and trustworthy.
Three Common Types of Constraints:
PRIMARY KEY:

Description: A PRIMARY KEY constraint uniquely identifies each record in a table. It must contain unique values, and it cannot contain NULL values. A table can only have one primary key, which can consist of one or more columns.
Scenario: In an Orders table, OrderID would be an excellent candidate for a primary key. Each order should have a unique ID, and this ID is essential for tracking and retrieving specific order details. If two orders had the same OrderID, it would be impossible to distinguish them.

CREATE TABLE Orders (
    OrderID INT PRIMARY KEY,
    CustomerID INT,
    OrderDate DATE,
    TotalAmount DECIMAL(10, 2)
);
FOREIGN KEY:

Description: A FOREIGN KEY is a field (or collection of fields) in one table that refers to the PRIMARY KEY in another table. It establishes a link between two tables, enforcing referential integrity, meaning that a value in the foreign key column must exist in the primary key column of the referenced table.
Scenario: Consider an Order_Items table that stores individual products within an order. Each item in Order_Items must be associated with an existing order in the Orders table. The OrderID in Order_Items would be a foreign key referencing the OrderID (primary key) in the Orders table. This prevents creating order items for non-existent orders.

CREATE TABLE Order_Items (
    OrderItemID INT PRIMARY KEY,
    OrderID INT,
    ProductID INT,
    Quantity INT,
    FOREIGN KEY (OrderID) REFERENCES Orders(OrderID)
);
NOT NULL:

Description: The NOT NULL constraint ensures that a column cannot have a NULL value. This means that a value must always be provided for that column when a new record is inserted or an existing record is updated.
Scenario: In a Customers table, columns like FirstName and LastName would typically have a NOT NULL constraint. It's usually a business requirement that every customer has a first and last name, so allowing NULL values would result in incomplete or unusable customer records.

CREATE TABLE Customers (
    CustomerID INT PRIMARY KEY,
    FirstName VARCHAR(50) NOT NULL,
    LastName VARCHAR(50) NOT NULL,
    Email VARCHAR(100)
);

QUESTION 3:

SOLUTION: LIMIT Clause
Purpose: The LIMIT clause is used to restrict the number of rows returned by a query. It's often used when you only need a specific number of records, such as the top N results, or for pagination.
Syntax: SELECT column1, column2 FROM table_name LIMIT number_of_rows;
Example: To get the first 5 employees from an Employees table:
SELECT * FROM Employees LIMIT 5;
OFFSET Clause
Purpose: The OFFSET clause is used to skip a specified number of rows before beginning to return the result rows. It's almost always used in conjunction with LIMIT for pagination, allowing you to fetch records starting from a particular point in the result set.
Syntax: SELECT column1, column2 FROM table_name LIMIT number_of_rows OFFSET start_row;
Example: To get records starting from the 6th row (skipping the first 5) and retrieve the next 5 records:
SELECT * FROM Employees LIMIT 5 OFFSET 5;
Using LIMIT and OFFSET for Pagination (Retrieving the third page of results with 10 records per page)
To retrieve the third page of results, assuming each page has 10 records, you would use both LIMIT and OFFSET as follows:

LIMIT: This will be 10 (the number of records per page).
OFFSET: To get to the third page, you need to skip the first two pages. Each page has 10 records, so you need to skip (page_number - 1) * records_per_page rows. For the third page, this would be (3 - 1) * 10 = 2 * 10 = 20 rows.
SELECT * FROM YourTable
LIMIT 10 OFFSET 20;
Explanation:

LIMIT 10: This tells the database to return a maximum of 10 rows.
OFFSET 20: This tells the database to skip the first 20 rows before starting to return the 10 rows specified by LIMIT.
Therefore, LIMIT 10 OFFSET 20 effectively retrieves rows 21 through 30, which constitutes the third page of results.

QUESTION 4:

SOLUTION: A Common Table Expression (CTE) is a temporary, named result set that you can reference within a single SQL statement (e.g., SELECT, INSERT, UPDATE, DELETE). It's defined using the WITH clause and is not permanently stored in the database. Think of it as a temporary view that exists only for the duration of the query.

Main Benefits of CTEs:
Readability and Maintainability: CTEs break down complex, multi-step queries into smaller, logical, and more readable blocks. This makes the SQL code easier to understand, debug, and maintain.

Simplifying Complex Joins and Subqueries: They can replace complex subqueries or derived tables, making the query structure much flatter and easier to follow, especially when dealing with multiple levels of nesting.

Recursion: CTEs are essential for performing recursive queries, which allow you to query hierarchical data (e.g., organizational charts, bill of materials) by repeatedly calling the CTE itself.

Reusability within a Single Query: A CTE can be referenced multiple times within the same SELECT, INSERT, UPDATE, or DELETE statement, avoiding redundant code and improving efficiency (though the database optimizer ultimately determines execution). This is particularly useful when you need to perform the same intermediate calculation or filtering step multiple times.

Simple SQL Example Demonstrating CTE Usage:
Let's imagine we have an Employees table with EmployeeID, FirstName, LastName, and Salary. We want to find employees who earn more than the average salary of all employees.

Without CTE (using a subquery):

SELECT FirstName, LastName, Salary
FROM Employees
WHERE Salary > (SELECT AVG(Salary) FROM Employees);
With CTE:

WITH AverageSalaryCTE AS (
    SELECT AVG(Salary) AS AvgSal
    FROM Employees
)
SELECT e.FirstName, e.LastName, e.Salary
FROM Employees e, AverageSalaryCTE avg_cte
WHERE e.Salary > avg_cte.AvgSal;
In this example, AverageSalaryCTE is the temporary result set that calculates the average salary. We then reference this CTE in the main SELECT statement to filter employees. While this specific example is simple enough that the benefit might not be immediately obvious, imagine if AverageSalaryCTE involved more complex calculations or joins; the CTE would significantly improve clarity.

QUESTION 5:

SOLUTION: SQL normalization is a database design technique primarily used to organize the columns and tables of a relational database to minimize data redundancy and improve data integrity. It involves decomposing larger tables into smaller, less redundant tables, and then defining relationships between these tables.

Primary Goals of Normalization:
Eliminate Data Redundancy: By storing data only once, normalization reduces the amount of storage space required and prevents inconsistencies that can arise when the same data is stored in multiple places and updated differently.
Improve Data Integrity: It ensures that data is consistent and reliable. Changes to data only need to be made in one place, reducing the chances of errors and maintaining accuracy.
Minimize Data Anomalies: Normalization helps to avoid insertion anomalies (difficulty adding new data without complete information), update anomalies (inconsistencies when updating redundant data), and deletion anomalies (unintended loss of data when deleting a record).
Simplify Querying: While sometimes requiring more joins, normalized databases can make queries more logical and consistent, especially for complex analytical tasks.
Enhance Database Scalability and Flexibility: A well-normalized database is generally easier to extend and maintain as business requirements change.
First Three Normal Forms (1NF, 2NF, 3NF):
First Normal Form (1NF):

Definition: A table is in 1NF if it meets the following criteria:
Each column contains atomic (indivisible) values. There are no multi-valued attributes in a single column.
Each column has a unique name.
The order of data does not matter.
There are no repeating groups of columns.
Simply put: "Each cell should contain only one value, and there should be no repeating groups of columns."
Example (Not in 1NF): A Customers table where a PhoneNumbers column contains multiple phone numbers separated by commas for a single customer. (To be in 1NF, each phone number would need its own row or be in a separate table).
Second Normal Form (2NF):

Definition: A table is in 2NF if it meets the following criteria:
It is in 1NF.
All non-key attributes are fully functionally dependent on the primary key. This means that every non-key attribute must depend on the entire primary key, not just a part of it (this applies to tables with composite primary keys).
Simply put: "It's in 1NF, and all non-key attributes depend on the whole primary key."
Example (Not in 2NF): An Order_Details table with a composite primary key (OrderID, ProductID). If ProductName depends only on ProductID (part of the primary key) and not on OrderID, then ProductName violates 2NF. ProductName should be moved to a separate Products table.
Third Normal Form (3NF):

Definition: A table is in 3NF if it meets the following criteria:
It is in 2NF.
There are no transitive dependencies. This means that a non-key attribute should not depend on another non-key attribute.
Simply put: "It's in 2NF, and no non-key attribute determines another non-key attribute."
Example (Not in 3NF): An Employees table with EmployeeID (PK), EmployeeName, DepartmentName, and DepartmentLocation. If DepartmentLocation depends on DepartmentName (a non-key attribute) rather than directly on EmployeeID, then DepartmentLocation violates 3NF. DepartmentName and DepartmentLocation should be moved to a separate Departments table.

QUESTION 6:

## Initialize Database and Connection



In [5]:
import sqlite3

# Establish a connection to an in-memory SQLite database
conn = sqlite3.connect(':memory:')

# Create a cursor object
cursor = conn.cursor()

print("In-memory SQLite database connection established successfully.")

In-memory SQLite database connection established successfully.


In [6]:
ddl_categories = """
CREATE TABLE Categories (
    CategoryID INTEGER PRIMARY KEY,
    CategoryName TEXT NOT NULL UNIQUE
);
"""

cursor.execute(ddl_categories)
conn.commit()

print("Table 'Categories' created successfully.")

Table 'Categories' created successfully.


## Create Products Table



In [7]:
ddl_products = """
CREATE TABLE Products (
    ProductID INTEGER PRIMARY KEY,
    ProductName TEXT NOT NULL UNIQUE,
    Price REAL NOT NULL,
    CategoryID INTEGER,
    FOREIGN KEY (CategoryID) REFERENCES Categories(CategoryID)
);
"""

cursor.execute(ddl_products)
conn.commit()

print("Table 'Products' created successfully.")

Table 'Products' created successfully.


## Create Customers Table



In [8]:
ddl_customers = """
CREATE TABLE Customers (
    CustomerID INTEGER PRIMARY KEY,
    FirstName TEXT NOT NULL,
    LastName TEXT NOT NULL,
    Email TEXT UNIQUE
);
"""

cursor.execute(ddl_customers)
conn.commit()

print("Table 'Customers' created successfully.")

Table 'Customers' created successfully.


## Create Orders Table



In [9]:
ddl_orders = """
CREATE TABLE Orders (
    OrderID INTEGER PRIMARY KEY,
    CustomerID INTEGER,
    OrderDate TEXT NOT NULL,
    TotalAmount REAL NOT NULL,
    FOREIGN KEY (CustomerID) REFERENCES Customers(CustomerID)
);
"""

cursor.execute(ddl_orders)
conn.commit()

print("Table 'Orders' created successfully.")

Table 'Orders' created successfully.


## Insert Sample Data into Categories Table



In [10]:
dml_categories = [
    (1, 'Electronics'),
    (2, 'Books'),
    (3, 'Home & Kitchen')
]

for category_id, category_name in dml_categories:
    cursor.execute("INSERT INTO Categories (CategoryID, CategoryName) VALUES (?, ?)", (category_id, category_name))

conn.commit()

print("Sample data inserted into 'Categories' table successfully.")

Sample data inserted into 'Categories' table successfully.


## Insert Sample Data into Products Table



In [11]:
dml_products = [
    (101, 'Laptop', 1200.00, 1),
    (102, 'Smartphone', 800.00, 1),
    (201, 'Python Programming', 45.50, 2),
    (202, 'Data Science Handbook', 60.00, 2),
    (301, 'Coffee Maker', 75.00, 3)
]

for product_id, product_name, price, category_id in dml_products:
    cursor.execute("INSERT INTO Products (ProductID, ProductName, Price, CategoryID) VALUES (?, ?, ?, ?)", (product_id, product_name, price, category_id))

conn.commit()

print("Sample data inserted into 'Products' table successfully.")

Sample data inserted into 'Products' table successfully.


## Insert Sample Data into Customers Table



In [12]:
dml_customers = [
    (1, 'Alice', 'Smith', 'alice.smith@example.com'),
    (2, 'Bob', 'Johnson', 'bob.johnson@example.com'),
    (3, 'Charlie', 'Brown', 'charlie.brown@example.com')
]

for customer_id, first_name, last_name, email in dml_customers:
    cursor.execute("INSERT INTO Customers (CustomerID, FirstName, LastName, Email) VALUES (?, ?, ?, ?)", (customer_id, first_name, last_name, email))

conn.commit()

print("Sample data inserted into 'Customers' table successfully.")

Sample data inserted into 'Customers' table successfully.


## Insert Sample Data into Orders Table



In [13]:
dml_orders = [
    (1001, 1, '2023-01-20', 1245.50),
    (1002, 2, '2023-01-22', 860.00),
    (1003, 1, '2023-01-25', 75.00),
    (1004, 3, '2023-01-28', 105.50)
]

for order_id, customer_id, order_date, total_amount in dml_orders:
    cursor.execute("INSERT INTO Orders (OrderID, CustomerID, OrderDate, TotalAmount) VALUES (?, ?, ?, ?)", (order_id, customer_id, order_date, total_amount))

conn.commit()

print("Sample data inserted into 'Orders' table successfully.")

Sample data inserted into 'Orders' table successfully.


## Verify Data Insertion



In [14]:
print("Verifying data in Categories table:")
cursor.execute("SELECT * FROM Categories")
categories_data = cursor.fetchall()
for row in categories_data:
    print(row)

Verifying data in Categories table:
(1, 'Electronics')
(2, 'Books')
(3, 'Home & Kitchen')


In [15]:
print("\nVerifying data in Products table:")
cursor.execute("SELECT * FROM Products")
products_data = cursor.fetchall()
for row in products_data:
    print(row)


Verifying data in Products table:
(101, 'Laptop', 1200.0, 1)
(102, 'Smartphone', 800.0, 1)
(201, 'Python Programming', 45.5, 2)
(202, 'Data Science Handbook', 60.0, 2)
(301, 'Coffee Maker', 75.0, 3)


In [16]:
print("\nVerifying data in Customers table:")
cursor.execute("SELECT * FROM Customers")
customers_data = cursor.fetchall()
for row in customers_data:
    print(row)


Verifying data in Customers table:
(1, 'Alice', 'Smith', 'alice.smith@example.com')
(2, 'Bob', 'Johnson', 'bob.johnson@example.com')
(3, 'Charlie', 'Brown', 'charlie.brown@example.com')


In [17]:
print("\nVerifying data in Orders table:")
cursor.execute("SELECT * FROM Orders")
orders_data = cursor.fetchall()
for row in orders_data:
    print(row)


Verifying data in Orders table:
(1001, 1, '2023-01-20', 1245.5)
(1002, 2, '2023-01-22', 860.0)
(1003, 1, '2023-01-25', 75.0)
(1004, 3, '2023-01-28', 105.5)


## Summary:

### Data Analysis Key Findings

*   An in-memory SQLite database connection was successfully established.
*   Four core e-commerce tables were successfully created using DDL statements:
    *   `Categories`: With `CategoryID` as primary key and `CategoryName` as a unique, non-null field.
    *   `Products`: Including `ProductID` as primary key, `ProductName` (unique, non-null), `Price` (non-null), and `CategoryID` as a foreign key.
    *   `Customers`: Featuring `CustomerID` as primary key, `FirstName` and `LastName` (non-null), and `Email` (unique).
    *   `Orders`: Comprising `OrderID` as primary key, `CustomerID` as a foreign key, `OrderDate` (non-null), and `TotalAmount` (non-null).
*   Sample data was successfully inserted into all tables using DML statements:
    *   3 records into the `Categories` table (e.g., 'Electronics', 'Books').
    *   5 records into the `Products` table (e.g., 'Laptop', 'Python Programming').
    *   3 records into the `Customers` table (e.g., 'Alice Smith', 'Bob Johnson').
    *   4 records into the `Orders` table (e.g., orders for customers 1, 2, and 3).
*   Data insertion was verified by executing DQL `SELECT *` statements for each table, confirming all records were present and correct.

### Insights or Next Steps

*   The successfully established and populated ECommerceDB database provides a foundational structure ready for complex querying and application development.
*   The next step could involve performing analytical queries across the linked tables (e.g., joining `Orders` and `Customers`, or `Products` and `Categories`) to extract business insights, or further extending the schema with additional tables like `OrderDetails` for more granular order information.


QUESTION 7:

##  Generate Customer Order Report

In [18]:
sql_query_q7 = """
SELECT
    c.FirstName || ' ' || c.LastName AS CustomerName,
    c.Email,
    COUNT(o.OrderID) AS TotalNumberofOrders
FROM
    Customers c
LEFT JOIN
    Orders o ON c.CustomerID = o.CustomerID
GROUP BY
    c.CustomerID, c.FirstName, c.LastName, c.Email
ORDER BY
    CustomerName;
"""

cursor.execute(sql_query_q7)
report_q7 = cursor.fetchall()

print("Customer Order Report:")
print("----------------------")
for row in report_q7:
    print(f"Name: {row[0]}, Email: {row[1]}, Total Orders: {row[2]}")

Customer Order Report:
----------------------
Name: Alice Smith, Email: alice.smith@example.com, Total Orders: 2
Name: Bob Johnson, Email: bob.johnson@example.com, Total Orders: 1
Name: Charlie Brown, Email: charlie.brown@example.com, Total Orders: 1


QUESTION 8:



##  Retrieve Product Information with Category

In [20]:
sql_query_q8 = """
SELECT
    p.ProductName,
    p.Price,
    c.CategoryName
FROM
    Products p
JOIN
    Categories c ON p.CategoryID = c.CategoryID
ORDER BY
    c.CategoryName, p.ProductName;
"""

cursor.execute(sql_query_q8)
products_with_categories = cursor.fetchall()

print("Product Information with Categories:")
print("------------------------------------어요")
for row in products_with_categories:
    print(f"Product: {row[0]}, Price: {row[1]}, Category: {row[2]}")

Product Information with Categories:
------------------------------------어요
Product: Data Science Handbook, Price: 60.0, Category: Books
Product: Python Programming, Price: 45.5, Category: Books
Product: Laptop, Price: 1200.0, Category: Electronics
Product: Smartphone, Price: 800.0, Category: Electronics
Product: Coffee Maker, Price: 75.0, Category: Home & Kitchen


QUESTION 9:


##  Top 2 Most Expensive Products per Category using CTE and Window Function

In [21]:
sql_query_q9 = """
WITH RankedProducts AS (
    SELECT
        c.CategoryName,
        p.ProductName,
        p.Price,
        ROW_NUMBER() OVER (PARTITION BY c.CategoryName ORDER BY p.Price DESC) as rn
    FROM
        Products p
    JOIN
        Categories c ON p.CategoryID = c.CategoryID
)
SELECT
    CategoryName,
    ProductName,
    Price
FROM
    RankedProducts
WHERE
    rn <= 2
ORDER BY
    CategoryName, Price DESC;
"""

cursor.execute(sql_query_q9)
ranked_products = cursor.fetchall()

print("Top 2 Most Expensive Products per Category:")
print("--------------------------------------------")
for row in ranked_products:
    print(f"Category: {row[0]}, Product: {row[1]}, Price: {row[2]}")

Top 2 Most Expensive Products per Category:
--------------------------------------------
Category: Books, Product: Data Science Handbook, Price: 60.0
Category: Books, Product: Python Programming, Price: 45.5
Category: Electronics, Product: Laptop, Price: 1200.0
Category: Electronics, Product: Smartphone, Price: 800.0
Category: Home & Kitchen, Product: Coffee Maker, Price: 75.0


QUESTION 10:


### Create Customer Table

In [22]:
ddl_customer = """
CREATE TABLE customer (
    customer_id INTEGER PRIMARY KEY,
    first_name TEXT NOT NULL,
    last_name TEXT NOT NULL,
    email TEXT UNIQUE,
    active INTEGER NOT NULL DEFAULT 1
);
"""

cursor.execute(ddl_customer)
conn.commit()

print("Table 'customer' created successfully.")

Table 'customer' created successfully.


### Insert Sample Data into Customer Table

In [23]:
dml_customer_data = [
    (1, 'John', 'Doe', 'john.doe@example.com', 1),
    (2, 'Jane', 'Smith', 'jane.smith@example.com', 1),
    (3, 'Peter', 'Jones', 'peter.jones@example.com', 0) # Inactive customer
]

for customer_id, first_name, last_name, email, active in dml_customer_data:
    cursor.execute("INSERT INTO customer (customer_id, first_name, last_name, email, active) VALUES (?, ?, ?, ?, ?)",
                   (customer_id, first_name, last_name, email, active))

conn.commit()

print("Sample data inserted into 'customer' table successfully.")

Sample data inserted into 'customer' table successfully.


### Create Category Table

In [24]:
ddl_category = """
CREATE TABLE category (
    category_id INTEGER PRIMARY KEY,
    name TEXT NOT NULL UNIQUE
);
"""

cursor.execute(ddl_category)
conn.commit()

print("Table 'category' created successfully.")

Table 'category' created successfully.


### Insert Sample Data into Category Table

In [25]:
dml_category_data = [
    (1, 'Action'),
    (2, 'Comedy'),
    (3, 'Drama'),
    (4, 'Sci-Fi')
]

for category_id, name in dml_category_data:
    cursor.execute("INSERT INTO category (category_id, name) VALUES (?, ?)",
                   (category_id, name))

conn.commit()

print("Sample data inserted into 'category' table successfully.")

Sample data inserted into 'category' table successfully.


### Create Film_Category Table

In [26]:
ddl_film_category = """
CREATE TABLE film_category (
    film_id INTEGER,
    category_id INTEGER,
    PRIMARY KEY (film_id, category_id),
    FOREIGN KEY (category_id) REFERENCES category(category_id)
);
"""

cursor.execute(ddl_film_category)
conn.commit()

print("Table 'film_category' created successfully.")

Table 'film_category' created successfully.


### Insert Sample Data into Film_Category Table

In [27]:
dml_film_category_data = [
    (1, 1), # Film 1 in Action
    (1, 2), # Film 1 also in Comedy
    (2, 1), # Film 2 in Action
    (3, 3), # Film 3 in Drama
    (4, 2), # Film 4 in Comedy
    (5, 4), # Film 5 in Sci-Fi
    (6, 1), # Film 6 in Action
    (6, 3)  # Film 6 also in Drama
]

for film_id, category_id in dml_film_category_data:
    cursor.execute("INSERT INTO film_category (film_id, category_id) VALUES (?, ?)",
                   (film_id, category_id))

conn.commit()

print("Sample data inserted into 'film_category' table successfully.")

Sample data inserted into 'film_category' table successfully.


### Create Inventory Table

In [28]:
ddl_inventory = """
CREATE TABLE inventory (
    inventory_id INTEGER PRIMARY KEY,
    film_id INTEGER NOT NULL,
    store_id INTEGER NOT NULL
);
"""

cursor.execute(ddl_inventory)
conn.commit()

print("Table 'inventory' created successfully.")

Table 'inventory' created successfully.


### Insert Sample Data into Inventory Table

In [29]:
dml_inventory_data = [
    (1, 1, 1), # Film 1 at Store 1
    (2, 1, 2), # Film 1 at Store 2
    (3, 2, 1), # Film 2 at Store 1
    (4, 3, 2), # Film 3 at Store 2
    (5, 4, 1), # Film 4 at Store 1
    (6, 5, 2), # Film 5 at Store 2
    (7, 6, 1), # Film 6 at Store 1
    (8, 6, 2)  # Film 6 at Store 2
]

for inventory_id, film_id, store_id in dml_inventory_data:
    cursor.execute("INSERT INTO inventory (inventory_id, film_id, store_id) VALUES (?, ?, ?)",
                   (inventory_id, film_id, store_id))

conn.commit()

print("Sample data inserted into 'inventory' table successfully.")

Sample data inserted into 'inventory' table successfully.


### Create Payment Table

In [30]:
ddl_payment = """
CREATE TABLE payment (
    payment_id INTEGER PRIMARY KEY,
    customer_id INTEGER NOT NULL,
    amount REAL NOT NULL,
    payment_date TEXT NOT NULL,
    FOREIGN KEY (customer_id) REFERENCES customer(customer_id)
);
"""

cursor.execute(ddl_payment)
conn.commit()

print("Table 'payment' created successfully.")

Table 'payment' created successfully.


In [31]:
dml_payment_data = [
    (1, 1, 19.99, '2023-01-20'),
    (2, 2, 25.50, '2023-01-21'),
    (3, 1, 12.00, '2023-01-22'),
    (4, 3, 30.00, '2023-01-23')
]

for payment_id, customer_id, amount, payment_date in dml_payment_data:
    cursor.execute("INSERT INTO payment (payment_id, customer_id, amount, payment_date) VALUES (?, ?, ?, ?)",
                   (payment_id, customer_id, amount, payment_date))

conn.commit()

print("Sample data inserted into 'payment' table successfully.")

Sample data inserted into 'payment' table successfully.


### Create Rental Table

In [32]:
ddl_rental = """
CREATE TABLE rental (
    rental_id INTEGER PRIMARY KEY,
    rental_date TEXT NOT NULL,
    inventory_id INTEGER NOT NULL,
    customer_id INTEGER NOT NULL,
    return_date TEXT,
    FOREIGN KEY (inventory_id) REFERENCES inventory(inventory_id),
    FOREIGN KEY (customer_id) REFERENCES customer(customer_id)
);
"""

cursor.execute(ddl_rental)
conn.commit()

print("Table 'rental' created successfully.")

Table 'rental' created successfully.


In [40]:
dml_rental_data = [
    (1, '2023-01-01', 1, 1, '2023-01-08'), # Customer 1 rents Film 1 from Store 1
    (2, '2023-01-02', 3, 2, '2023-01-09'), # Customer 2 rents Film 2 from Store 1
    (3, '2023-01-03', 2, 1, '2023-01-10'), # Customer 1 rents Film 1 from Store 2
    (4, '2023-01-04', 4, 3, None),         # Customer 3 rents Film 3 from Store 2 (not returned yet)
    (5, '2023-01-05', 7, 1, '2023-01-12')  # Customer 1 rents Film 6 from Store 1
]

for rental_id, rental_date, inventory_id, customer_id, return_date in dml_rental_data:
    cursor.execute("INSERT INTO rental (rental_id, rental_date, inventory_id, customer_id, return_date) VALUES (?, ?, ?, ?, ?)",
                   (rental_id, rental_date, inventory_id, customer_id, return_date))

conn.commit()

print("Sample data inserted into 'rental' table successfully.")

IntegrityError: UNIQUE constraint failed: rental.rental_id

In [34]:
dml_rental_data = [
    (1, '2023-01-01', 1, 1, '2023-01-08'), # Customer 1 rents Film 1 from Store 1
    (2, '2023-01-02', 3, 2, '2023-01-09'), # Customer 2 rents Film 2 from Store 1
    (3, '2023-01-03', 2, 1, '2023-01-10'), # Customer 1 rents Film 1 from Store 2
    (4, '2023-01-04', 4, 3, None),         # Customer 3 rents Film 3 from Store 2 (not returned yet)
    (5, '2023-01-05', 7, 1, '2023-01-12')  # Customer 1 rents Film 6 from Store 1
]

for rental_id, rental_date, inventory_id, customer_id, return_date in dml_rental_data:
    cursor.execute("INSERT INTO rental (rental_id, rental_date, inventory_id, customer_id, return_date) VALUES (?, ?, ?, ?, ?)",
                   (rental_id, rental_date, inventory_id, customer_id, return_date))

conn.commit()

print("Sample data inserted into 'rental' table successfully.")

Sample data inserted into 'rental' table successfully.


In [35]:
sql_query_q10 = """
SELECT
    c.first_name || ' ' || c.last_name AS CustomerName,
    c.email,
    SUM(p.amount) AS TotalSpent
FROM
    customer c
JOIN
    payment p ON c.customer_id = p.customer_id
GROUP BY
    c.customer_id, c.first_name, c.last_name, c.email
ORDER BY
    TotalSpent DESC
LIMIT 5;
"""

cursor.execute(sql_query_q10)
top_customers = cursor.fetchall()

print("Top 5 Customers by Total Spent:")
print("--------------------------------")
for row in top_customers:
    print(f"Name: {row[0]}, Email: {row[1]}, Total Spent: {row[2]:.2f}")

Top 5 Customers by Total Spent:
--------------------------------
Name: John Doe, Email: john.doe@example.com, Total Spent: 31.99
Name: Peter Jones, Email: peter.jones@example.com, Total Spent: 30.00
Name: Jane Smith, Email: jane.smith@example.com, Total Spent: 25.50


In [36]:
sql_query_q11 = """
SELECT
    c.name AS CategoryName,
    COUNT(r.rental_id) AS RentalCount
FROM
    category c
JOIN
    film_category fc ON c.category_id = fc.category_id
JOIN
    inventory i ON fc.film_id = i.film_id
JOIN
    rental r ON i.inventory_id = r.inventory_id
GROUP BY
    c.name
ORDER BY
    RentalCount DESC
LIMIT 3;
"""

cursor.execute(sql_query_q11)
top_categories = cursor.fetchall()

print("Top 3 Movie Categories by Rental Count:")
print("---------------------------------------")
for row in top_categories:
    print(f"Category: {row[0]}, Rental Count: {row[1]}")

Top 3 Movie Categories by Rental Count:
---------------------------------------
Category: Action, Rental Count: 4
Category: Drama, Rental Count: 2
Category: Comedy, Rental Count: 2


## Calculate Film Availability and Unrented Films per Store



In [37]:
sql_query_q12 = """
SELECT
    i.store_id,
    COUNT(i.inventory_id) AS TotalFilmsAvailable,
    SUM(CASE WHEN r.rental_id IS NULL THEN 1 ELSE 0 END) AS UnrentedFilms
FROM
    inventory i
LEFT JOIN
    rental r ON i.inventory_id = r.inventory_id
GROUP BY
    i.store_id
ORDER BY
    i.store_id;
"""

cursor.execute(sql_query_q12)
film_availability_report = cursor.fetchall()

print("Film Availability and Unrented Films per Store:")
print("--------------------------------------------------")
for row in film_availability_report:
    print(f"Store ID: {row[0]}, Total Films: {row[1]}, Unrented Films: {row[2]}")

Film Availability and Unrented Films per Store:
--------------------------------------------------
Store ID: 1, Total Films: 4, Unrented Films: 1
Store ID: 2, Total Films: 4, Unrented Films: 2


## Analyze Monthly Revenue for 2023



In [38]:
sql_query_q13 = """
SELECT
    strftime('%m', payment_date) AS PaymentMonth,
    SUM(amount) AS TotalRevenue
FROM
    payment
WHERE
    strftime('%Y', payment_date) = '2023'
GROUP BY
    PaymentMonth
ORDER BY
    PaymentMonth;
"""

cursor.execute(sql_query_q13)
monthly_revenue_2023 = cursor.fetchall()

print("Monthly Revenue for 2023:")
print("-------------------------")
for row in monthly_revenue_2023:
    print(f"Month: {row[0]}, Total Revenue: {row[1]:.2f}")

Monthly Revenue for 2023:
-------------------------
Month: 01, Total Revenue: 87.49


## Identify Frequent Renters in Last 6 Months



In [39]:
sql_query_q14 = """
SELECT
    c.first_name || ' ' || c.last_name AS CustomerName,
    c.email,
    COUNT(r.rental_id) AS RentalCount
FROM
    customer c
JOIN
    rental r ON c.customer_id = r.customer_id
WHERE
    r.rental_date >= DATE('2023-07-01', '-6 months')
GROUP BY
    c.customer_id, c.first_name, c.last_name, c.email
HAVING
    RentalCount > 10
ORDER BY
    RentalCount DESC;
"""

cursor.execute(sql_query_q14)
frequent_renters = cursor.fetchall()

print("Frequent Renters (Last 6 Months, >10 Rentals):")
print("--------------------------------------------------")

if frequent_renters:
    for row in frequent_renters:
        print(f"Name: {row[0]}, Email: {row[1]}, Rental Count: {row[2]}")
else:
    print("No customers found who rented more than 10 times in the last 6 months.")
    print("(Note: This is expected given the small sample data size, which is designed for demonstrating query logic.)")


Frequent Renters (Last 6 Months, >10 Rentals):
--------------------------------------------------
No customers found who rented more than 10 times in the last 6 months.
(Note: This is expected given the small sample data size, which is designed for demonstrating query logic.)


## Final Task




## Summary:

### Q&A

*   **Who are the top 5 customers by total spending?**
    The top customers by total spending are: John Doe with \$31.99, Peter Jones with \$30.00, and Jane Smith with \$25.50. (Only three customers had payment records in the sample data).

*   **Which are the top 3 movie categories by rental count?**
    The top 3 movie categories by rental count are: Action with 4 rentals, Drama with 2 rentals, and Comedy with 2 rentals.

*   **What is the film availability and number of unrented films per store?**
    *   Store ID 1 has 4 total films, with 1 unrented film.
    *   Store ID 2 has 4 total films, with 2 unrented films.

*   **What was the monthly revenue for 2023?**
    For January 2023, the total revenue generated was \$87.49. (Only January had payment records in the sample data).

*   **Were any customers identified as frequent renters (more than 10 rentals in the last 6 months)?**
    No customers met the criteria of more than 10 rentals in the last 6 months, which is an expected outcome given the small size of the sample database.

### Data Analysis Key Findings

*   **Database Setup:** A minimal Sakila database was successfully set up with `customer`, `payment`, `category`, `film_category`, `inventory`, and `rental` tables, populated with sufficient sample data for the analysis.
*   **Customer Spending:** The highest spending customer in the sample data was John Doe, with a total expenditure of \$31.99, followed by Peter Jones (\$30.00) and Jane Smith (\$25.50).
*   **Popular Categories:** The 'Action' movie category was the most rented with 4 rental transactions, followed by 'Drama' and 'Comedy' categories, both with 2 rentals each.
*   **Film Availability:** Both Store ID 1 and Store ID 2 have 4 films in their inventory. Store ID 1 has 1 unrented film, while Store ID 2 has 2 unrented films.
*   **Monthly Revenue:** The sample data showed revenue generation only in January 2023, totaling \$87.49.
*   **Renter Activity:** No customers in the sample dataset qualified as "frequent renters" (more than 10 rentals in the last 6 months), indicating the sample data is primarily for demonstrating query functionality rather than reflecting real-world high-volume activity.

### Insights or Next Steps

*   The sample data, while sufficient for demonstrating SQL query functionality, is too small to draw meaningful business insights regarding customer behavior, category popularity, or revenue seasonality.
*   To perform a more robust analysis, the next step should involve populating the database with a larger and more diverse dataset that reflects realistic rental patterns and customer activity over a longer period.
